### PR TITLE
Add generic CRC algorithm

### PR DESCRIPTION
**Pull request by erikcorry**
Replaces the nibble-at-a-time CRC32 algorithm with a byte-at-a-time generic algorithm.
Replaces the bit-at-a-time CRC16 algorithm with a byte-at-a-time generic algorithm.

### CODE REVIEW

File: lib/crypto/crc16.toit [hunk 1]
```diff
@@ -17,40 +18,11 @@ The $data must be a string or byte array.
...
-  /**
-  See $super.
-
-  Returns the CRC16 checksum as a 2 element byte array in little-endian order.
-  */
-  get -> ByteArray:
-    checksum := sum_
-    return ByteArray 2: (checksum >> (8 * it)) & 0xff
+class Crc16 extends Crc:
+  constructor:
```
Comment (floitsch) [inline lib/crypto/crc16.toit:27]: I would make the polynomial a default value of a named parameter.

Maybe even deprecate a constructor without any polynomial.

Not completely sure how to do it, but the Modem Crc16 is not *that* common to make it the default.

File: lib/crypto/crc32.toit [hunk 2]
```diff
@@ -14,26 +14,17 @@ The $data must be a string or byte array.
 Returns the checksum as a 4 element byte array in little-endian order.
 */
 crc32 data from/int=0 to/int=data.size -> ByteArray:
```
Comment (floitsch) [inline lib/crypto/crc32.toit:16]: I would make the initial-state and xor_result be optional parameters. (unless nobody uses any else).

For crc32, the function might be good enough, as there aren't many different implementations.

File: lib/crypto/crc32.toit [hunk 3]
```diff
@@ -14,26 +14,17 @@ The $data must be a string or byte array.
...
-
-  /**
-  See $super.
-
-  Returns the CRC32 checksum as a 4 element byte array in little-endian order.
-  */
-  get -> ByteArray:
-    checksum := sum_ ^ 0xffffffff
-    return ByteArray 4: (checksum >> (8 * it)) & 0xff
+    super.little_endian 32
```
Comment (floitsch) [inline lib/crypto/crc32.toit:27]: At least for the constructor I would probably allow different polynomials, initial_state and/or xor_results.

**Review by floitsch — APPROVED** [review]
LGTM.

File: lib/crypto/crc.toit [hunk 4]
```diff
@@ -1,43 +1,223 @@
...
+    (reversed) ordering for the polynomial.
+
+  # Example
+  ```
+  // The popular CRC-32 used in PNG.
+  crc := Crc.little_endian 32 --polynomial=0xEDB88320 --initial_state=0xffff_ffff --xor_result=0xffff_ffff
+  ```
+  */
+  constructor.little_endian .width/int --.polynomial/int --initial_state/int=0 --.xor_result/int=0:
+    if polynomial > 1 << width: throw "Polynomial and width don't match"
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply lib/crypto/crc.toit:52]: I think the () are noise here.

**Comment by erikcorry** [comment]
I think I want to deprecate Crc16 completely and replace it with classes and methods that use the name from crccalc.com, which is CRC-16/XMODEM, which would be Crc16Xmodem.

For Crc32 that is the name used at crccalc.com, so I would keep it, but not add config options, since using those would make it a different algorithm that did not match the name any more.

But I'll do this in a new change.

File: lib/crypto/crc.toit (hunk 4 above)
[the remark this answers is not in the document]
Comment (floitsch) [reply lib/crypto/crc.toit:52]: I disagree. I misread it the first time.
The `>` and `<<` look too similar in this context.

**Pull request merged**